test/extended/baremetal/helper.go

```diff
@@ -0,0 +1,198 @@
+
+	var secret v1.Secret
+	err = yaml.Unmarshal(yamlData, &secret)
+	o.Expect(err).ToNot(o.HaveOccurred())
+	return &secret
+}
+
+// getExtraWorkerSecretData gets and decodes the baremetal host associated for the
+// specified extra worker. If not found, the test will fail
+func (b *BaremetalTestHelper) getExtraWorkerData(index int) *unstructured.Unstructured {
+	ew, err := b.extraWorkersRetrieveData()
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	yamlData, ok := ew.Data[b.extraWorkerKey(index)]
+	o.Expect(ok).To(o.BeTrue(), fmt.Sprintf("unable to find data for extra worker %d", index))
+
+	jsonData, err := yaml.YAMLToJSON(yamlData)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	var host unstructured.Unstructured
+	err = host.UnmarshalJSON(jsonData)
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	return &host
+}
+
+// GetExtraWorkerData gets baremetal host and secret for the specified extra worker.
+// If not found, the test will fail
+func (b *BaremetalTestHelper) GetExtraWorkerData(index int) (*unstructured.Unstructured, *v1.Secret) {
+	return b.getExtraWorkerData(index), b.getExtraWorkerSecretData(index)
+}
+
+// CreateExtraWorker creates a new BaremetalHost (and associated Secret).
+// If successfull, returns the newly created host and secret resources
+func (b *BaremetalTestHelper) CreateExtraWorker(host *unstructured.Unstructured, secret *v1.Secret) (*unstructured.Unstructured, *v1.Secret) {
+
+	secret, err := b.clientSet.CoreV1().Secrets("openshift-machine-api").Create(context.Background(), secret, metav1.CreateOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	host, err = b.bmcClient.Create(context.Background(), host, metav1.CreateOptions{})
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	b.hostsCreated = append(b.hostsCreated, host)
+
+	return host, secret
+}
+
+// DeleteAllExtraWorkers deletes all the extra workers created in the current session
+func (b *BaremetalTestHelper) DeleteAllExtraWorkers() {
+	for _, host := range b.hostsCreated {
+		hostName := host.GetName()
+
+		err := b.bmcClient.Delete(context.Background(), hostName, metav1.DeleteOptions{})
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		b.waitForDeletion(host)
+	}
+}
+
+func (b *BaremetalTestHelper) waitForDeletion(host *unstructured.Unstructured) {
+	err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (done bool, err error) {
+		_, err = b.bmcClient.Get(context.TODO(), host.GetName(), metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return true, nil
+			}
+
+			return false, err
+		}
+		return false, nil
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+// WaitForProvisioningState waits for the given baremetal host to reach the specified provisioning state.
+// If successfull, returns the updated host resource, otherwise the test will fail
+func (b *BaremetalTestHelper) WaitForProvisioningState(host *unstructured.Unstructured, expectedProvisioningState string) *unstructured.Unstructured {
```

Comment:
It looks like there is no point that calling `CreateExtraWorker` and not calling this function. Maybe, we can use it in `CreateExtraWorker` instead giving call responsibility to test owner?

Reply:
I didn't want to overengineer the API of the helper too much for the moment, and considered that the WaitForProvisioningState could be a more generic function useful for other tests (and other scenarios). I expect we're going to refine the interface as long as we're using it in more tests

Reply:
Added `DeployExtraWorker`